completer/completer.7
.TH YED-COMPLETER 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
completer \- A basic completion plugin activated on TAB.
.SH CONFIGURATION
.SS completer-sources [ sources... ]
Space-delimited list of completion sources to use. Default is "word".
.SS completer-popup [ yes / no ]
Whether or not do display the popup menu for completion results.
Default value is "yes".
.SS completer-popup-size [ int ]
The size in lines of the popup menu.
 Default value is 8.
.SS completer-auto [ yes / no ]
Whether or not the popup should come up automatically as you type in an "autocomplete" fashion.
Default value is "no".
.SH COMMANDS
None
.SH BUFFERS
None
.SH NOTES
.P
Completer uses the word ([a-Z0-9_]+) before the cursor as the search string.
.SH VERSION
0.0.1
.SH KEYWORDS
complete, completion, auto, autocomplete, popup, tab, TAB


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


plugins/completer/completer.c
#include <yed/plugin.h>

typedef struct {
    yed_frame *frame;
    array_t    strings;
    array_t    dds;
    int        start_len;
    int        start_x;
    int        is_up;
    int        selection;
} completer_popup_t;

static completer_popup_t popup;

char *completer_get_current_word(void);
void completer_buff_pre_insert_handler(yed_event *event);
void completer_buff_post_insert_handler(yed_event *event);
void completer_buff_post_delete_back_handler(yed_event *event);
void completer_cursor_post_move_handler(yed_event *event);
void completer_frame_activated_handler(yed_event *event);

int yed_plugin_boot(yed_plugin *self) {
    yed_event_handler h;

    YED_PLUG_VERSION_CHECK();

    if (yed_get_var("completer-sources") == NULL) {
        yed_set_var("completer-sources", "word");
    }

    if (yed_get_var("completer-popup") == NULL) {
        yed_set_var("completer-popup", "yes");
    }

    if (yed_get_var("completer-auto") == NULL) {
        yed_set_var("completer-auto", "no");
    }

    if (yed_get_var("completer-popup-size") == NULL) {
        yed_set_var("completer-popup-size", "8");
    }


    h.kind = EVENT_BUFFER_PRE_INSERT;
    h.fn   = completer_buff_pre_insert_handler;
    yed_plugin_add_event_handler(self, h);

    h.kind = EVENT_BUFFER_POST_INSERT;
    h.fn   = completer_buff_post_insert_handler;
    yed_plugin_add_event_handler(self, h);

    h.kind = EVENT_BUFFER_POST_DELETE_BACK;
    h.fn   = completer_buff_post_delete_back_handler;
    yed_plugin_add_event_handler(self, h);

    h.kind = EVENT_CURSOR_POST_MOVE;
    h.fn   = completer_cursor_post_move_handler;
    yed_plugin_add_event_handler(self, h);

    h.kind = EVENT_FRAME_ACTIVATED;
    h.fn   = completer_frame_activated_handler;
    yed_plugin_add_event_handler(self, h);

    return 0;
}

static void completer_kill_popup(void) {
    yed_direct_draw_t **dd;

    if (!popup.is_up) { return; }

    free_string_array(popup.strings);

    array_traverse(popup.dds, dd) {
        yed_kill_direct_draw(*dd);
    }

    array_free(popup.dds);

    popup.frame = NULL;

    popup.is_up = 0;
}

static void completer_draw_popup(void) {
    yed_direct_draw_t **dd_it;
    yed_attrs           active;
    yed_attrs           assoc;
    yed_attrs           merged;
    yed_attrs           merged_inv;
    char              **it;
    int                 max_width;
    int                 has_left_space;
    int                 size;
    int                 start_idx;
    int                 i;
    char                buff[512];
    yed_direct_draw_t  *dd;

    array_traverse(popup.dds, dd_it) {
        yed_kill_direct_draw(*dd_it);
    }
    array_free(popup.dds);

    popup.dds = array_make(yed_direct_draw_t*);

    active = yed_active_style_get_active();
    assoc  = yed_active_style_get_associate();
    merged = active;
    yed_combine_attrs(&merged, &assoc);
    merged_inv = merged;
    merged_inv.flags ^= ATTR_INVERSE;

    max_width = strlen("--MORE--");
    array_traverse(popup.strings, it) {
        max_width = MAX(max_width, strlen(*it));
    }

    has_left_space = popup.start_x > 1;

    size = 8;
    yed_get_var_as_int("completer-popup-size", &size);

    start_idx = MAX(0, popup.selection - size + 1);
    i         = 0;

    array_traverse_from(popup.strings, it, start_idx) {
        snprintf(buff, sizeof(buff), "%s%*s ", has_left_space ? " " : "", -max_width, *it);
        dd = yed_direct_draw(popup.frame->cur_y + i + 1,
                             popup.start_x - has_left_space,
                             start_idx + i == popup.selection ? merged_inv : merged,
                             buff);
        array_push(popup.dds, dd);

        i += 1;
        if (i == size) {
            if (start_idx + i == array_len(popup.strings)) {
                snprintf(buff, sizeof(buff), "%s%*s ", has_left_space ? " " : "", -max_width, "--END--");
            } else {
                snprintf(buff, sizeof(buff), "%s%*s ", has_left_space ? " " : "", -max_width, "--MORE--");
            }
            dd = yed_direct_draw(popup.frame->cur_y + i + 1,
                                 popup.start_x - has_left_space,
                                 merged,
                                 buff);
            array_push(popup.dds, dd);
            break;
        }
    }
}

static void completer_start_popup(yed_frame *frame, int start_len, array_t strings) {
    completer_kill_popup();

    popup.frame     = frame;
    popup.strings   = copy_string_array(strings);
    popup.dds       = array_make(yed_direct_draw_t*);
    popup.start_len = start_len;
    popup.start_x   = frame->cur_x - start_len;
    popup.selection = -1;

    completer_draw_popup();

    popup.is_up = 1;
}

char *completer_get_current_word(void) {
    yed_frame *frame;
    yed_line  *line;
    char       c;
    char       prev_c;

    frame = ys->active_frame;

    if (frame == NULL || frame->buffer == NULL) { return NULL; }

    line = yed_buff_get_line(frame->buffer, frame->cursor_line);
    if (line == NULL) { return NULL; }

    if (frame->cursor_col <= 1) { return NULL; }

    c = prev_c = 0;

    if (frame->cursor_col <= line->visual_width) {
        c = ((yed_glyph*)yed_line_col_to_glyph(line, frame->cursor_col))->c;
    }

    prev_c = ((yed_glyph*)yed_line_col_to_glyph(line, frame->cursor_col - 1))->c;

    /* There must be part of a word behind the cursor. */
    if (!isalnum(prev_c) && prev_c != '_') { return NULL; }

    /* The cursor must NOT be on part of a word. */
    if (isalnum(c) || c == '_') { return NULL; }

    return yed_word_at_point(frame, frame->cursor_line, frame->cursor_col - 1);
}

static int do_complete(char *word, yed_completion_results *results) {
    int      status;
    char    *sources_str;
    array_t  sources;
    char    *tok;

    status = COMPL_ERR_NO_ERR;

    sources_str = yed_get_var("completer-sources");
    if (sources_str == NULL) { return COMPL_ERR_NO_COMPL; }

    sources_str = strdup(sources_str);
    sources     = array_make(char*);

    for (tok = strtok(sources_str, " "); tok != NULL; tok = strtok(NULL, " ")) {
        array_push(sources, tok);
    }

    if (array_len(sources) > 0) {
        if (array_len(sources) == 1) {
            status = yed_complete(*(char**)array_data(sources), word, results);
        } else {
            status = yed_complete_multiple(array_len(sources), array_data(sources), word, results);
        }
    }

    array_free(sources);
    free(sources_str);

    return status;
}

int completer_cmd_line(void) {
    yed_frame               *frame;
    yed_buffer              *buff;
    char                    *word;
    int                      word_len;
    int                      compl_status;
    yed_completion_results   results;
    int                      compl_num_items;
    int                      cpl;
    char                    *first;
    int                      i;
    char                   **it;

    frame = ys->active_frame;
    if (frame == NULL) { return 0; }

    buff = frame->buffer;
    if (buff == NULL) { return 0; }

    word = completer_get_current_word();
    if (word == NULL) { return 0; }

    word_len = strlen(word);

    compl_status    = do_complete(word, &results);
    compl_num_items = array_len(results.strings);
    cpl             = results.common_prefix_len;

    if (compl_status == COMPL_ERR_NO_ERR && compl_num_items > 0) {
        first = *(char**)array_item(results.strings, 0);

        yed_start_undo_record(frame, buff);
        for (i = word_len; i < cpl; i += 1) {
            yed_insert_into_line(buff, frame->cursor_line, frame->cursor_col, G(first[i]));
            yed_move_cursor_within_frame(frame, 0, 1);
        }
        yed_end_undo_record(frame, buff);

        if (compl_num_items > 1) {
LOG_CMD_ENTER("completer");
            yed_cprint("");
            array_traverse(results.strings, it) {
                yed_cprint("\n%s", *it);
            }
LOG_EXIT();
        }

        free_string_array(results.strings);
    }

    free(word);

    return 1;
}

static int completer_no_auto_popup(void) {
    yed_frame               *frame;
    yed_buffer              *buff;
    char                    *word;
    int                      word_len;
    int                      compl_status;
    yed_completion_results   results;
    int                      compl_num_items;
    int                      cpl;
    char                    *first;
    int                      i;

    frame = ys->active_frame;
    if (frame == NULL) { return 0; }

    buff = frame->buffer;
    if (buff == NULL) { return 0; }

    word = completer_get_current_word();
    if (word == NULL) { return 0; }

    word_len = strlen(word);

    compl_status    = do_complete(word, &results);
    compl_num_items = array_len(results.strings);
    cpl             = results.common_prefix_len;

    if (compl_status == COMPL_ERR_NO_ERR && compl_num_items > 0) {
        first = *(char**)array_item(results.strings, 0);

        yed_start_undo_record(frame, buff);
        for (i = word_len; i < cpl; i += 1) {
            yed_insert_into_line(buff, frame->cursor_line, frame->cursor_col, G(first[i]));
            yed_move_cursor_within_frame(frame, 0, 1);
        }
        yed_end_undo_record(frame, buff);

        if (compl_num_items > 1) {
            completer_start_popup(frame, cpl, results.strings);
        }

        free_string_array(results.strings);
    }

    free(word);

    return 1;
}

static int completer_auto_popup(void) {
    yed_frame               *frame;
    yed_buffer              *buff;
    char                    *word;
    int                      word_len;
    int                      compl_status;
    yed_completion_results   results;
    int                      compl_num_items;

    frame = ys->active_frame;
    if (frame == NULL) { return 0; }

    buff = frame->buffer;
    if (buff == NULL) { return 0; }

    word = completer_get_current_word();
    if (word == NULL) { return 0; }

    word_len = strlen(word);

    compl_status    = do_complete(word, &results);
    compl_num_items = array_len(results.strings);

    if (compl_status == COMPL_ERR_NO_ERR && compl_num_items > 0) {
        completer_start_popup(frame, word_len, results.strings);
        free_string_array(results.strings);
    }

    free(word);

    return 1;
}

static void completer_next_selection(void) {
    yed_frame *frame;
    char      *sel_string;
    int        i;

    frame = popup.frame;

    yed_start_undo_record(frame, frame->buffer);
    if (popup.selection > -1) {
        sel_string = *(char**)array_item(popup.strings, popup.selection);
        for (i = 0; i < strlen(sel_string) - popup.start_len; i += 1) {
            yed_move_cursor_within_frame(frame, 0, -1);
            yed_delete_from_line(frame->buffer, frame->cursor_line, frame->cursor_col);
        }
    }

    popup.selection += 1;
    if (popup.selection == array_len(popup.strings)) {
        popup.selection = -1;
    }

    if (popup.selection > -1) {
        sel_string = *(char**)array_item(popup.strings, popup.selection);
        for (i = popup.start_len; i < strlen(sel_string); i += 1) {
            yed_insert_into_line(frame->buffer, frame->cursor_line, frame->cursor_col, G(sel_string[i]));
            yed_move_cursor_within_frame(frame, 0, 1);
        }
    }

    yed_end_undo_record(frame, frame->buffer);

    completer_draw_popup();
}

static void completer_prev_selection(void) {
    yed_frame *frame;
    char      *sel_string;
    int        i;

    frame = popup.frame;

    yed_start_undo_record(frame, frame->buffer);
    if (popup.selection > -1) {
        sel_string = *(char**)array_item(popup.strings, popup.selection);
        for (i = 0; i < strlen(sel_string) - popup.start_len; i += 1) {
            yed_move_cursor_within_frame(frame, 0, -1);
            yed_delete_from_line(frame->buffer, frame->cursor_line, frame->cursor_col);
        }
    }

    popup.selection -= 1;
    if (popup.selection == -2) {
        popup.selection = array_len(popup.strings) - 1;
    }

    if (popup.selection > -1) {
        sel_string = *(char**)array_item(popup.strings, popup.selection);
        for (i = popup.start_len; i < strlen(sel_string); i += 1) {
            yed_insert_into_line(frame->buffer, frame->cursor_line, frame->cursor_col, G(sel_string[i]));
            yed_move_cursor_within_frame(frame, 0, 1);
        }
    }

    yed_end_undo_record(frame, frame->buffer);

    completer_draw_popup();
}

static int in_pre_insert_handler;
static int in_post_insert_handler;

void completer_buff_pre_insert_handler(yed_event *event) {
    int did_compl;

    if (in_pre_insert_handler || in_post_insert_handler) { return; }

    if (event->key == ENTER) {
        if (popup.is_up && popup.selection > -1) {
            completer_kill_popup();
            event->cancel = 1;
            return;
        }
    }

    if (event->key != TAB && event->key != SHIFT_TAB) { return; }

    in_pre_insert_handler = 1;

    did_compl = 0;

    if (yed_var_is_truthy("completer-popup")) {
        if (popup.is_up) {
            if (event->key == TAB) {
                completer_next_selection();
            } else if (event->key == SHIFT_TAB) {
                completer_prev_selection();
            }
            did_compl = 1;
        } else if (yed_var_is_truthy("completer-auto")) {
            did_compl = completer_auto_popup();
        } else {
            did_compl = completer_no_auto_popup();
        }
    } else {
        did_compl = completer_cmd_line();
    }

    if (did_compl) {
        event->cancel = 1;
    }

    in_pre_insert_handler = 0;
}

void completer_buff_post_insert_handler(yed_event *event) {
    if (!yed_var_is_truthy("completer-auto")
    ||  !yed_var_is_truthy("completer-popup"))           { return; }
    if (in_pre_insert_handler || in_post_insert_handler) { return; }

    in_post_insert_handler = 1;

    completer_kill_popup();
    completer_auto_popup();

    in_post_insert_handler = 0;
}

void completer_buff_post_delete_back_handler(yed_event *event) {
    if (!yed_var_is_truthy("completer-auto")
    ||  !yed_var_is_truthy("completer-popup"))           { return; }
    if (in_pre_insert_handler || in_post_insert_handler) { return; }

    in_post_insert_handler = 1;

    completer_kill_popup();
    completer_auto_popup();

    in_post_insert_handler = 0;
}

void completer_cursor_post_move_handler(yed_event *event) {
    if (in_pre_insert_handler || in_post_insert_handler) { return; }

    if (popup.is_up && event->frame == popup.frame) {
        completer_kill_popup();
    }
}

void completer_frame_activated_handler(yed_event *event) {
    if (in_pre_insert_handler || in_post_insert_handler) { return; }

    if (popup.is_up) {
        completer_kill_popup();
    }
}
